File: hash/js/rosetta-image-hash/tests/group1-median-filter.test.ts
import { describe, it, expect } from "vitest";
import { readFileSync } from "node:fs";
import { join } from "node:path";
import { SPEC_DIR } from "./testkit.js";
import { pilMedianFilter } from "../src/internal/pilMedianFilter.js";

interface MedianFilterCase {
  name: string;
  shape: [number, number];
  input: number[][];
  output: number[][];
}

interface MedianFilterCases {
  cases: MedianFilterCase[];
}

describe("pilMedianFilter", () => {
  it("byte-exact against all median_filter_cases.json cases", () => {
    const path = join(SPEC_DIR, "median_filter_cases.json");
    const doc = JSON.parse(readFileSync(path, "utf8")) as MedianFilterCases;
    expect(doc.cases.length).toBeGreaterThan(0);

    const failures: string[] = [];
    for (const c of doc.cases) {
      const [height, width] = c.shape;
      const flat = new Uint8Array(height * width);
      for (let y = 0; y < height; y++) {
        for (let x = 0; x < width; x++) {
          flat[y * width + x] = c.input[y][x];
        }
      }
      const got = pilMedianFilter(flat, width, height);
      for (let y = 0; y < height; y++) {
        for (let x = 0; x < width; x++) {
          const idx = y * width + x;
          if (got[idx] !== c.output[y][x]) {
            failures.push(
              `${c.name} (${y},${x}): got ${got[idx]} want ${c.output[y][x]}`
            );
          }
        }
      }
    }
    if (failures.length > 0) {
      throw new Error(`${failures.length} pixel failures:\n  ${failures.slice(0, 10).join("\n  ")}`);
    }
  });
});

File: hash/rust/rosetta-image-hash/tests/group1_median_filter_test.rs
// Test-only suppressions: the for-loop indices mirror the reference
// Python loops; clippy's iterator suggestion would obscure the port
// equivalence and is not load-bearing for test code.
#![allow(clippy::needless_range_loop)]

mod testkit;

use rosetta_image_hash::internal::pil_median_filter;
use serde::Deserialize;
use std::fs;
use testkit::spec_path::SPEC_DIR;

#[derive(Debug, Deserialize)]
struct MedianFilterCase {
    name: String,
    input: Vec<Vec<u8>>,
    output: Vec<Vec<u8>>,
}

#[derive(Debug, Deserialize)]
struct MedianFilterCases {
    cases: Vec<MedianFilterCase>,
}

fn load_cases() -> MedianFilterCases {
    let path = format!("{}/median_filter_cases.json", SPEC_DIR);
    let data = fs::read_to_string(&path).unwrap_or_else(|e| panic!("read {path}: {e}"));
    serde_json::from_str(&data).expect("parse median_filter_cases.json")
}

#[test]
fn median_filter_group1() {
    let cases = load_cases();
    let mut failures: Vec<String> = Vec::new();

    for c in &cases.cases {
        let got = pil_median_filter::apply(&c.input);

        if got != c.output {
            let mut first_diff = String::new();
            'outer: for y in 0..c.output.len() {
                for x in 0..c.output[y].len() {
                    if got[y][x] != c.output[y][x] {
                        first_diff = format!(
                            "first diff at ({},{}) got={} want={}",
                            y, x, got[y][x], c.output[y][x]
                        );
                        break 'outer;
                    }
                }
            }
            failures.push(format!("case={} {}", c.name, first_diff));
        }
    }

    if !failures.is_empty() {
        panic!("{} median_filter failures:\n  {}", failures.len(), failures.join("\n  "));
    }
}
